// results/index.js
var RequestController = require('express').Router({mergeParams: true}),
SQLStatements = require('sql-statements'),
PGService = require('pg-service-c'),
ProjectResultsModel = require('results/ProjectResultsModel.js');

RequestController
.route('/')
.get(function (request, response) {
  if (!request.params.id) {
    return response.status(400).end();
  }
  var project = {
    projectID: Number(request.params.id)
  },
  results = [];
  for (var i = 0; i < ProjectResultsModel.length; i++) {
    if (ProjectResultsModel[i].projectID === project.projectID) {
      results.push(ProjectResultsModel[i]);
    }
  }
  return response.status(200)
  .json(results).end();
})
.post(function (request, response) {
  console.log(request.params.id);
  if (!request.params.id) {
    return response.status(400).end();
  }
  PGService.insertRecord({
    text: SQLStatements['project results']['ru']['insert project result'],
    values: [
      request.params.id,
      request.body.typeID,
      request.body.resultName,
      request.body.parentID,
      request.body.authorID,
      request.body.responsiblePersonID
    ]
  }, function (status, data) {
    if (status === 201) {
      ProjectResultsModel.push({
        id: data.id,
        projectID: request.params.id,
        typeID: request.body.typeID,
        resultName: request.body.resultName,
        parentID: request.body.parentID,
        authorID: request.body.authorID,
        responsiblePersonID: request.body.responsiblePersonID,
        isDeleted: 'N'
      });
    }
    return response.status(status)
    .json(data).end();
  });
})

module.exports = RequestController;
