// companies/index.js
var RequestController = require('express').Router(),
RequestValidator = require('request-validator'),
SQLStatements = require('sql-statements'),
PGService = require('pg-service-c'),
CompaniesModel = require('companies/CompaniesModel.js');

/* @RequestController */
RequestController
/** @function get companies count */
.head('/', function (request, response) {
  PGService.selectRecords({
    text: SQLStatements['companies']['ru']['select companies count']
  }, function (status, data) {
    return response.status(200)
    .append('Companies-Count', data[0].count).end();
  });
})
/** @function get companies */
.get('/', function (request, response) {
  if (CompaniesModel.length === 0) {
    return response.status(204)
    .json([]).end();
  } else {
    return response.status(200)
    .json(CompaniesModel).end();
  }
})
/** @function get person */
.get('/:id', function (request, response) {
  var bin = request.params.id,
  res = RequestValidator.validate(bin, 'iin');
  if (!res.result) {
    return response.status(400)
    .json({
      error: res.message
    }).end();
  }
  for (var i = 0, len = CompaniesModel.length; i < len; i++) {
    if (CompaniesModel[i].id === bin) {
      return response.status(200)
      .json(CompaniesModel[i]).end();
    }
  }
  return response.status(204)
  .json([]).end();
})
/** @function create person */
.post('/', function (request, response) {
  var bin = request.body.id;
  res = RequestValidator.validate(bin, 'iin');
  if (!res.result) {
    return response.status(400)
    .json({
      error: res.message
    }).end();
  }
  for (var i = 0; i < CompaniesModel.length; i++) {
    if (CompaniesModel[i].id === bin) {
      return response.status(400)
      .json({
        error: 'company with '+bin+' exist'
      }).end();
    }
  }
  PGService.updateRecord({
    text: SQLStatements['companies']['ru']['insert company'],
    values: [
      bin,
      request.body.shortName,
      request.body.longName,
      request.body.fullName
    ]
  }, function (status, data) {
    if ((status === 201 && bin === data.id)) {
      CompaniesModel.push({
        id: bin,
        shortName: request.body.shortName,
        longName: request.body.longName,
        fullName: request.body.fullName,
        isDeleted: 'N'
      });
    }
    return response.status(status)
    .json(data).end();
  });
})
/** @function update person */
.put('/:id', function (request, response) {
  var bin = request.params.id,
  res = RequestValidator.validate(bin, 'iin')
  if (!res.result) {
    return response.status(400)
    .json({
      error: res.message
    }).end();
  }
  PGService.updateRecord({
    text: SQLStatements['companies']['ru']['update company'],
    values: [
      bin,
      request.body.shortName,
      request.body.longName,
      request.body.fullName
    ]
  }, function (status, data) {
    if (status === 200) {
      for (var i = 0; i < CompaniesModel.length; i++) {
        if (CompaniesModel[i].id === data.id) {
          CompaniesModel[i].shortName = request.body.shortName,
          CompaniesModel[i].longName = request.body.longName,
          CompaniesModel[i].fullName = request.body.fullName
        }
      }
    }
    return response.status(status)
    .json(data).end();
  });
})
/** @function delete person */
.delete('/:id', function (request, response) {
  var bin = request.params.id,
  res = RequestValidator.validate(bin, 'iin')
  if (!res.result) {
    return response.status(400)
    .json({
      error: res.message
    }).end();
  }
  PGService.updateRecord({
    text: SQLStatements['companies']['ru']['delete company'],
    values: [request.params.id]
  }, function (status, data) {
    if (status === 200) {
      for (var i = 0; i < CompaniesModel.length; i++) {
        if (CompaniesModel[i].id === data.id) {
          CompaniesModel[i].isDeleted = 'Y'
        }
      }
    }
    return response.status(status)
    .json(data).end();
  });
})
/** @function restore person */
.options('/:id', function (request, response) {
  var bin = request.params.id,
  res = RequestValidator.validate(bin, 'iin');
  if (!res.result) {
    return response.status(400)
    .json({
      error: res.message
    }).end();
  }
  PGService.updateRecord({
    text: SQLStatements['companies']['ru']['restore company'],
    values: [request.params.id]
  }, function (status, data) {
    if (status === 200) {
      for (var i = 0; i < CompaniesModel.length; i++) {
        if (CompaniesModel[i].id === data.id) {
          CompaniesModel[i].isDeleted = 'N'
        }
      }
    }
    return response.status(status)
    .json(data).end();
  });
});

module.exports = RequestController;
