// routes/projects.js
'use strict';

var ProjectsRouter = require('express').Router(),
DBQueries = require('../db/DBQueries.js'),
Controller = require('../controller'), PGController = new Controller(),
DBService = require('../db/DBService.js'), PGService = new DBService();
/* RequestController */
ProjectsRouter
/* RequestService */
.get('/', function (request, response) {
  PGService.performQueryWL(
    DBQueries.selectProjectsQuery(),
    PGService.selectRecordsWL,
    function (status, data) {
      return response.status(status)
      .json(data).end();
    }
  );
})
.get('/:id', function (request, response) {
  if (!request.params.id) {
    return response.status(400).end();
  }
  var project = {
    id: Number(request.params.id)
  };
  if (PGController.validate(project, 'id')) {
    PGService.performQueryWL(
      DBQueries.selectProjectQuery(project),
      PGService.selectRecordWL,
      function (status, data) {
        return response.status(status)
        .json(data).end();
      }
    );
  } else {
    return response.status(400).end();
  }
})
.post('/', function (request, response) {
  var project = {
    /* required length: 12 */
    customerID: String(request.data.customerID),
    /* required length: 2 */
    projectFormalName: String(request.data.projectFormalName),
    /* required length: 1 - 300 */
    projectWorkName: String(request.data.projectWorkName),
    /* length: 1 - 500 */
    projectOfficialName: String(request.data.projectOfficialName),
    startDate: new Date(request.data.startDate),
    endDate: new Date(request.data.endDate),
    budget: Number(request.data.budget),
    /* length: 12 */
    managerID: String(request.data.managerID)
  };
  if (PGController.validate(project, 'projectData')) {
    PGService.performQueryWL(
      DBQueries.selectProjectQuery(project),
      PGService.selectRecordWL,
      function (status, data) {
        return response.status(status)
        .json(data).end();
      }
    );
  } else {
    return response.status(400).end();
  }
})
.put('/:id', function (request, response) {
  if (!request.params.id) {
    return response.status(400).end();
  }
  var project = {
    /* required */
    id: Number(request.params.id),
    /* required length: 12 */
    customerID: String(request.data.customerID),
    /* required length: 2 */
    projectFormalName: String(request.data.projectFormalName),
    /* required length: 1 - 300 */
    projectWorkName: String(request.data.projectWorkName),
    /* length: 1 - 500 */
    projectOfficialName: String(request.data.projectOfficialName),
    startDate: new Date(request.data.startDate),
    endDate: new Date(request.data.endDate),
    budget: Number(request.data.budget),
    /* length: 12 */
    managerID: String(request.data.managerID)
  };
  if (PGController.validate(project, 'project')) {
    PGService.performQueryWL(
      DBQueries.updateProjectQuery(project),
      PGService.updateRecordWL,
      function (status, data) {
        return response.status(status)
        .json(data).end();
      }
    );
  } else {
    return response.status(400).end();
  }
})
.delete('/:id', function (request, response) {
  if (!request.params.id) {
    return response.status(400).end();
  }
  var project = {
    id: Number(request.params.id)
  };
  if (PGController.validate(project, 'id')) {
    PGService.performQueryWL(
      DBQueries.deleteProjectQuery(project),
      PGService.deleteRecordWL,
      function (status, data) {
        return response.status(status)
        .json(data).end();
      }
    );
  } else {
    return response.status(400).end();
  }
})

module.exports = ProjectsRouter;
