// db/DBQueries.js
var DBQueries = {
  selectTasksQuery: function selectTasksQuery() {
    return {
      text: 'SELECT id, task_name, task_description FROM e_tasks ORDER BY iin ASC'
    }
  },
  selectTaskQuery: function selectTasksQuery(task) {
    return {
      text: 'SELECT id, task_name, task_description FROM e_tasks WHERE id = {id}',
      values: task
    };
  },
  insertTaskQuery: function insertTaskQuery(task) {
    return {
      text: 'INSERT INTO e_tasks (task_name, task_description) VALUES ({taskName}, {taskDescription}) RETURNING id',
      values: task
    }
  },
  updateTaskQuery: function updateTaskQuery(task) {
    return {
      text: 'UPDATE e_tasks SET task_name = {taskName}, task_description = {taskDescription} WHERE id = {id}',
      values: task
    }
  },
  deleteTaskQuery: function deleteTaskQuery(task) {
    return {
      text: 'DELETE FROM e_tasks WHERE id = {id}',
      values: task
    }
  },
  selectPersonsQuery: function selectPersonsQuery() {
    return {
      text: 'SELECT iin, last_name AS "lastName", first_name AS "firstName", middle_name AS "middleName", dob, gender_id AS "genderID", is_deleted AS "isDeleted" FROM e_persons ORDER BY iin ASC'
    }
  },
  selectPersonQuery: function selectPersonQuery(person) {
    return {
      text: 'SELECT iin, last_name AS "lastName", first_name AS "firstName", middle_name AS "middleName", dob, gender_id AS "genderID", is_deleted AS "isDeleted" FROM e_persons WHERE iin = {iin}',
      values: person
    }
  }
};

module.exports = DBQueries;
