// db/DBQueries.js
var DBQueries = {
  selectProjectsQuery: function selectProjectsQuery() {
    return {
      text: 'SELECT id, customer_id AS "customerID", project_formal_name AS "projectFormalName", project_work_name AS "projectWorkName", project_official_name AS "projectOfficialName", start_date AS "startDate", end_date AS "endDate", budget, manager_id AS "managerID" FROM open_project.e_projects WHERE is_deleted = \'N\' ORDER BY id ASC'
    }
  },
  selectProjectQuery: function selectProjectQuery(project) {
    return {
      text: 'SELECT id, customer_id AS "customerID", project_formal_name AS "projectFormalName", project_work_name AS "projectWorkName", project_official_name AS "projectOfficialName", start_date AS "startDate", end_date AS "endDate", budget, manager_id AS "managerID" FROM open_project.e_projects WHERE is_deleted = \'N\' AND id = {id}',
      values: project
    }
  },
  insertProjectQuery: function insertProjectQuery(project) {
    return {
      text: 'INSERT INTO open_project.e_projects (customer_id, project_formal_name, project_work_name, project_official_name, start_date, end_date, budget, manager_id) VALUES ({customerID}, {projectFormalName}, {projectWorkName}, {projectOfficialName}, {startDate}, {endDate}, {budget}, {managerID}) RETURNING id',
      values: project
    }
  },
  updateProjectQuery: function updateProjectQuery(project) {
    return {
      text: 'UPDATE open_project.e_projects SET customer_id = {customerID}, project_formal_name = {projectFormalName}, project_work_name = {projectWorkName}, project_official_name = {projectOfficialName}, start_date = {startDate}, end_date = {endDate}, budget = {budget}, manager_id = {managerID} WHERE id = {id} RETURNING id',
      values: project
    }
  },
  deleteProjectQuery: function deleteProjectQuery(project) {
    return {
      text: 'DELETE FROM open_project.e_projects WHERE id = {id}',
      values: project
    }
  },
  selectGendersQuery: function selectGendersQuery() {
    return {
      text: 'SELECT id, gender_name AS "genderName" FROM fl.gender ORDER BY id DESC'
    }
  },
  selectTasksQuery: function selectTasksQuery() {
    return {
      text: 'SELECT id, task_name, task_description FROM e_tasks ORDER BY iin ASC'
    }
  },
  selectTaskQuery: function selectTasksQuery(task) {
    return {
      text: 'SELECT id, task_name, task_description FROM e_tasks WHERE id = {id}',
      values: task
    };
  },
  insertTaskQuery: function insertTaskQuery(task) {
    return {
      text: 'INSERT INTO e_tasks (task_name, task_description) VALUES ({taskName}, {taskDescription}) RETURNING id',
      values: task
    }
  },
  updateTaskQuery: function updateTaskQuery(task) {
    return {
      text: 'UPDATE e_tasks SET task_name = {taskName}, task_description = {taskDescription} WHERE id = {id}',
      values: task
    }
  },
  deleteTaskQuery: function deleteTaskQuery(task) {
    return {
      text: 'DELETE FROM e_tasks WHERE id = {id}',
      values: task
    }
  },
  selectPersonsQuery: function selectPersonsQuery() {
    return {
      text: 'SELECT id, last_name AS "lastName", first_name AS "firstName", middle_name AS "middleName", dob, gender_id AS "genderID" FROM fl.e_persons WHERE is_deleted = \'N\' ORDER BY id ASC'
    }
  },
  selectPersonsQueryLog: function selectPersonsQueryLog(personLog) {
    return {
      text: 'INSERT INTO log$.e_persons (session_id, manipulation_type_id) VALUES ({sessionID}, {manipulationTypeID}) RETURNING id',
      values: personLog
    }
  },
  selectPersonQuery: function selectPersonQuery(person) {
    return {
      text: 'SELECT id, last_name AS "lastName", first_name AS "firstName", middle_name AS "middleName", dob, gender_id AS "genderID" FROM fl.e_persons WHERE is_deleted = \'N\' AND id = {id}',
      values: person
    }
  },
  insertPersonQuery: function insertPersonQuery(person) {
    return {
    text: 'INSERT INTO fl.e_persons (id, last_name, first_name, middle_name, dob, gender_id) VALUES ({id}, {lastName}, {firstName}, {middleName}, {dob}, {genderID}) RETURNING id',
    values: person
    }
  },
  insertPersonQueryLog: function insertPersonQueryLog(personLog) {
    return {
    text: 'INSERT INTO log.e_persons (session_id, manipulation_type_id, iin, last_name, first_name, middle_name, dob, gender_id) values ({sessionID}, {manipulationTypeID}, {iin}, {lastName}, {firstName}, {middleName}, {dob}, {genderID}) RETURNING id',
    values: personLog
    }
  },
  updatePersonQuery: function updatePersonQuery(person) {
    return {
      text: 'UPDATE e_persons SET id = {id}, last_name = {lastName}, first_name = {firstName}, middle_name = {middleName}, dob = {dob}, gender_id = {genderID} WHERE id = {id} RETURNING id',
      values: person
    }
  },
  deletePersonQuery: function deletePersonQuery(person) {
    return {
      text: 'UPDATE e_persons SET is_deleted = 1 WHERE id = {id} RETURNING id',
      values: person
    }
  },
  selectCompaniesQuery: function selectCompaniesQuery() {
    return {
      text: 'SELECT id, short_name AS "shortName", long_name AS "longName", full_name AS "fullName" FROM ul.e_companies WHERE is_deleted = \'N\' ORDER BY id ASC'
    }
  },
  selectCompaniesQueryLog: function selectCompaniesQueryLog(companyLog) {
    return {
      text: 'INSERT INTO log$.e_persons (session_id, manipulation_type_id) VALUES ({sessionID}, {manipulationTypeID}) RETURNING id'
    }
  },
  insertCompanyQuery: function insertCompanyQuery(company) {
    return {
      text: 'INSERT INTO e_companies (id, company_name) VALUES ({id}, {companyName}) RETURNING id',
      values: company
    }
  },
  selectUserQuery: function selectUserQuery(user) {
    return {
      text: "SELECT u_password FROM meta.e_users WHERE is_blocked = 'N' AND is_deleted = 'N' AND u_username = {username}",
      values: user
    }
  }
};

module.exports = DBQueries;
