// pg-service-c/index.js
'use strict';

var pg = require('pg');

var PGService = {
  selectRecords: function selectRecords(query, cb) {
    if (!(query && typeof query === 'object')) {
      throw new Error('query not found');
      return cb(500, 'error');
    }
    if (!(query.text && typeof query.text === 'string')) {
      throw new Error('query text not found');
      return cb(500, 'error');
    }
    pg.connect(function (err, client, done) {
      if (err) {
        throw new Error(err);
        return cb(500, 'error');
      } else {
        client.query({
          text: query.text
        }, function (err, result) {
          var rowCount;
          if (err) {
            throw new Error(err);
            return cb(500, 'error');
          } else {
            done();
            rowCount = result.rowCount;
            if (rowCount === 0) {
              return cb(204, []);
            } else if (rowCount > 0) {
              return cb(200, result.rows);
            }
          }
        });
      }
    });
  },
  selectRecord: function selectRecord(query, cb) {
    if (!(query && typeof query === 'object')) {
      throw new Error('query not found');
      return cb(500, 'error');
    }
    if (!(query.text && typeof query.text === 'string')) {
      throw new Error('query text not found');
      return cb(500, 'error');
    }
    if (!(query.values && typeof query.values === 'object')) {
      throw new Error('query values not found');
      return cb(500, 'error');
    }
    if (!(query.values.id && typeof query.values.id === 'number')) {
      console.log('id not found');
      return cb(400, 'error');
    }
    pg.connect(function (err, client, done) {
      if (err) {
        throw new Error(err);
        return cb(500, 'error');
      } else {
        client.query({
          text: query.text,
          values: query.values
        }, function (err, result) {
          var rowCount;
          if (err) {
            throw new Error(err);
            return cb(500, 'error');
          } else {
            done();
            rowCount = result.rowCount;
            if (rowCount === 0) {
              return cb(204, []);
            } else if (rowCount === 1) {
              return cb(200, result.rows[0]);
            }
          }
        });
      }
    });
  },
  selectRecordsByID: function selectRecords(query, cb) {
    if (!(query && typeof query === 'object')) {
      throw new Error('query not found');
      return cb(500, 'error');
    }
    if (!(query.text && typeof query.text === 'string')) {
      throw new Error('query text not found');
      return cb(500, 'error');
    }
    if (!(query.values && typeof query.values === 'object')) {
      throw new Error('query values not found');
      return cb(500, 'error');
    }
    if (!(query.values.id && typeof query.values.id === 'number')) {
      console.log('id not found');
      return cb(400, 'error');
    }
    pg.connect(function (err, client, done) {
      if (err) {
        throw new Error(err);
        return cb(500, 'error');
      } else {
        client.query({
          text: query.text,
          values: query.values
        }, function (err, result) {
          var rowCount;
          if (err) {
            throw new Error(err);
            return cb(500, 'error');
          } else {
            done();
            rowCount = result.rowCount;
            if (rowCount === 0) {
              return cb(204, []);
            } else if (rowCount > 0) {
              return cb(200, result.rows);
            }
          }
        });
      }
    });
  },
  insertRecord: function insertRecord(query, cb) {
    if (!(query && typeof query === 'object')) {
      throw new Error('query not found');
      return cb(500, 'error');
    }
    if (!(query.text && typeof query.text === 'string')) {
      throw new Error('query text not found');
      return cb(500, 'error');
    }
    if (!(query.values && typeof query.values === 'object')) {
      throw new Error('query values not found');
      return cb(500, 'error');
    }
    pg.connect(function (err, client, done) {
      if (err) {
        throw new Error(err);
        return cb(500, 'error');
      } else {
        client.query({
          text: query.text,
          values: query.values
        }, function (err, result) {
          var rowCount, status;
          if (err) {
            if (err.code === '23505') {
              return cb(400, err.detail);
            }
            throw new Error(err);
            return cb(500, 'error');
          } else {
            done();
            rowCount = result.rowCount;
            if (rowCount === 0) {
              return cb(204, []);
            } else if (rowCount === 1) {
              return cb(201, result.rows[0]);
            }
          }
        });
      }
    });
  },
  updateRecord: function updateRecord(query, cb) {
    if (!(query && typeof query === 'object')) {
      throw new Error('query not found');
      return cb(500, 'error');
    }
    if (!(query.text && typeof query.text === 'string')) {
      throw new Error('query text not found');
      return cb(500, 'error');
    }
    if (!(query.values && typeof query.values === 'object')) {
      throw new Error('query values not found');
      return cb(500, 'error');
    }
    if (!(query.values.id && typeof query.values.id === 'number')) {
      console.log('id not found');
      return cb(400, 'error');
    }
    pg.connect(function (err, client, done) {
      if (err) {
        throw new Error(err);
        return cb(500, 'error');
      } else {
        client.query({
          text: query.text,
          values: query.values
        }, function (err, result) {
          var rowCount, status;
          if (err) {
            if (err.code === '23505') {
              return cb(400, err.detail);
            }
            throw new Error(err);
            return cb(500, 'error');
          } else {
            done();
            rowCount = result.rowCount;
            if (rowCount === 0) {
              return cb(204, []);
            } else if (rowCount === 1) {
              return cb(200, result.rows[0]);
            }
          }
        });
      }
    });
  }
};

module.exports = PGService;
